spec/controllers/text_components_controller_spec.rb

```diff
@@ -68,6 +68,41 @@
     end
   end
 
+  describe "GET #duplicate" do
+    let(:question_answer) { create(:question_answer, question: 'Question?', answer: 'Answer!') }
+    let(:text_component) { create(:text_component, report_id: report.id, heading: 'Heading', question_answer_ids: [question_answer.id]) }
+    before do
+      get :duplicate, params: {report_id: report.id, id: text_component.id}, session: valid_session
+    end
+
+    it "creates a new text component from an existing one" do
+      expect(assigns(:text_component)).not_to eq(text_component)
+    end
+
+    it "is does not persist the new text component" do
+      expect(assigns(:text_component)).not_to be_persisted
+    end
+
+    it "appends 'COPY' to the text component's heading" do
+      expect(assigns(:text_component).heading).to eq('Heading COPY')
+    end
+
+    it "resets the duplicate's publication status to 'draft'" do
+      expect(assigns(:text_component).draft?).to be_truthy
+    end
+
+    it "duplicates questions and answers as well" do
+      text_component.question_answers.each.with_index do |qa, i|
+        expect(assigns(:text_component).question_answers[i].question).to eq(qa.question)
+        expect(assigns(:text_component).question_answers[i].answer).to eq(qa.answer)
+      end
+    end
+
+    it "copys the text component's channels" do
+      expect(assigns(:text_component).channels).to match_array(text_component.channels)
+    end
+  end
+
```

Comment:
Great! I would also love to see a test that includes a `POST` request. Maybe a cucumber test. I would like to make sure that the entire `TextComponent` and it's association is untouched by the duplicate.

Reply:
Hm, okay. The duplicate action renders the new view which in turn posts to the new action, so I thought this wouldn’t be necessary. The `dup` method duplicates a record without it’s associations so any associations need to be set explicitly. But you’re right, I could probably add a test to make sure that questions and answers and triggers will always be duplicated or, respectively, removed.

Reply:
I don't know this behaviour. Just write some tests to show it, so other developers can see the desired, explicit behavior. If it is very low-level, a model test (unit test) might be enough already

Reply:
Okay, I extended the controller test to make sure that the associations are duplicated as well. I started writing a cucumber test as well, but had to deal with timing issues I wasn’t able to fix in a reasonable time. What do you think, are these tests sufficient.